### ISSUE TITLE
Add support for recursive neural networks

### ISSUE THREAD

**Issue by ibab** [tensorflow/tensorflow#2237]
I'd like to implement a recursive neural network as in [[Socher et al. 2011]](http://www.socher.org/index.php/Main/ParsingNaturalScenesAndNaturalLanguageWithRecursiveNeuralNetworks) in TensorFlow.

Note that this is different from recurrent neural networks, which are nicely supported by TensorFlow. The difference is that the network is not replicated into a linear sequence of operations, but into a tree structure.

Maybe a tree traversal can already be implemented using `While` ops, but I'm not sure how.
(A bottom-up traversal of a tree would have to be performed for each entry in the dataset, applying the network at each node)
Instinctively, I have the impression that this would need a new C++ op that generalizes `While`.

**Comment by zheng-xq** [comment]
Adding @yuanbyu to comment on how to use the control flow the implement recursive network.

My guess is that the control flow plus stack might be helpful.

**Comment by ibab** [comment]
I think a "stack of tensors" data structure could be implemented in a similar way to the existing [`TensorArray`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/tensor_array_ops.py#L36). I'm not sure how to propagate the gradients, though.

**Comment by ibab** [comment]
Looks like it's possible to implement this in Theano: https://github.com/ofirnachum/tree_rnn
(Haven't taken a closer look at the implementation yet, though)

**Comment by girving** [comment]
@yuanbyu: Do you have any suggestions?  I'll mark this contributions welcome for now.  If it's added, we should make sure it supports efficient batching.

**Comment by ibab** [comment]
I think the hardest part about this is coming up with a nice way of storing tree-shaped entries and passing them to Tensorflow ops.
Then a monolithic tensorflow op could provide endpoints for a tensor computation, which allows the user to define the network that will be replicated at each node of the tree.
The API could be something like this:
```python
def network(x, y):
  # Define transformation
  return z

# Evaluation
tf.recursive_eval(network, "tree_data_input", "tree_data_output")
# Training
loss = tf.recursive_loss(network, "tree_data_input", "tree_data_reference")
```

**Comment by kingtaurus** [comment]
This is an NLP style model (sentiment analysis) and is based upon assignment 3 from [Stanford CS224D](http://cs224d.stanford.edu/syllabus.html). The tree structure is created as a purely python structure (tree class and node class). The variable state is saved/restored on a per batch usage and the tree structure is created using recursive descent (with any leaf nodes using an embedding operation to generate a feature vector), the loss could be created in a similar manner.

The tree class is made up of nodes (tree structure is created through a parser - I'm leaving the details out for now):
```
class Node:  # a node in the tree
    def __init__(self, label, word=None):
        self.label = label
        self.word = word
        self.parent = None  # reference to parent
        self.left = None  # reference to left child
        self.right = None  # reference to right child
        # true if I am a leaf (could have probably derived this from if I have
        # a word)
        self.isLeaf = False
        # true if we have finished performing fowardprop on this node (note,
        # there are many ways to implement the recursion.. some might not
        # require this flag)
```
The tensorflow contributions, there are few parts:
(1) Variable creation (**NOTE**: `saver.restore(sess, './weights/%s.temp'%self.config.model_name)` is used to reload state) :
```
    def add_model_vars(self):
        '''
        You model contains the following parameters:
            embedding:  tensor(vocab_size, embed_size)
            W1:         tensor(2* embed_size, embed_size)
            b1:         tensor(1, embed_size)
            U:          tensor(embed_size, output_size)
            bs:         tensor(1, output_size)
        Hint: Add the tensorflow variables to the graph here and *reuse* them while building
                the compution graphs for composition and projection for each tree
        Hint: Use a variable_scope "Composition" for the composition layer, and
              "Projection") for the linear transformations preceding the softmax.
        '''
        with tf.variable_scope('Composition'):
            ### YOUR CODE HERE
            #initializer=initializer=tf.random_normal_initializer(0,3)
            embedding = tf.get_variable("embedding",
                                        [self.vocab.total_words, self.config.embed_size])
            W1 = tf.get_variable("W1", [2 * self.config.embed_size, self.config.embed_size])
            b1 = tf.get_variable("b1", [1, self.config.embed_size])
            ### END YOUR CODE
        with tf.variable_scope('Projection'):
            ### YOUR CODE HERE
            U = tf.get_variable("U", [self.config.embed_size, self.config.label_size])
            bs = tf.get_variable("bs", [1, self.config.label_size])
            ### END YOUR CODE
```
(2) model construction:
```
def add_model(self, node):
        """Recursively build the model to compute the phrase embeddings in the tree

        Hint: Refer to tree.py and vocab.py before you start. Refer to
              the model's vocab with self.vocab
        Hint: Reuse the "Composition" variable_scope here
        Hint: Store a node's vector representation in node.tensor so it can be
              used by it's parent
        Hint: If node is a leaf node, it's vector representation is just that of the
              word vector (see tf.gather()).
        Args:
            node: a Node object
        Returns:
            node_tensors: Dict: key = Node, value = tensor(1, embed_size)
        """
        with tf.variable_scope('Composition', reuse=True):
            ### YOUR CODE HERE
            embedding = tf.get_variable("embedding")
            W1 = tf.get_variable("W1")
            b1 = tf.get_variable("b1")
            l2_loss = tf.nn.l2_loss(W1) + tf.nn.l2_loss(b1)
            tf.add_to_collection(name="l2_loss", value=l2_loss)
            ### END YOUR CODE

        W_split = tf.split(0, 2, W1)
        W_left  = W_split[0]
        W_right = W_split[1]
        node_tensors = OrderedDict()
        curr_node_tensor = None
        if node.isLeaf:
            ### YOUR CODE HERE
            word_id = self.vocab.encode(node.word)
            curr_node_tensor = tf.expand_dims(tf.gather(embedding, word_id),0)
            ### END YOUR CODE
        else:
            node_tensors.update(self.add_model(node.left))
            node_tensors.update(self.add_model(node.right))
            ### YOUR CODE HERE
            #This operation could be done without the split call above
            #curr_node_tensor = tf.nn.relu(tf.matmul(child_tensor, W1) + b1)
            curr_node_tensor = tf.matmul(node_tensors[node.left], W_left) + tf.matmul(node_tensors[node.right], W_right) + b1
            ### END YOUR CODE
        node_tensors[node] = curr_node_tensor
        return node_tensors
```
(3) the projection layer is used to construct the sentiment for the root node:
```
    def add_projections(self, node_tensors):
        """Add projections to the composition vectors to compute the raw sentiment scores

        Hint: Reuse the "Projection" variable_scope here
        Args:
            node_tensors: tensor(?, embed_size)
        Returns:
            output: tensor(?, label_size)
        """
        logits = None
        ### YOUR CODE HERE
        with tf.variable_scope("Projection", reuse=True):
            U = tf.get_variable("U")
            bs = tf.get_variable("bs")
        logits = tf.matmul(node_tensors, U) + bs
        ### END YOUR CODE
        return logits
```

(4) loss (note: that you can use all the nodes, or just the root node for inference):
 `tf.reduce_sum(tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=labels))`,
where `logits` is given by a call to `inference`

```
    def inference(self, tree, predict_only_root=False):
        """For a given tree build the RNN models computation graph up to where it
            may be used for inference.
        Args:
            tree: a Tree object on which to build the computation graph for the RNN
        Returns:
            softmax_linear: Output tensor with the computed logits.
        """
        node_tensors = self.add_model(tree.root)
        if predict_only_root:
            node_tensors = node_tensors[tree.root]
        else:
            node_tensors = [tensor for node, tensor in node_tensors.items() if node.label!=2]
            node_tensors = tf.concat(0, node_tensors)
        return self.add_projections(node_tensors)
```

**Comment by marcellomh** [comment]
Hi Folks,

I'm a Google employee.  Our team has already implemented a library that handles recursive neural nets.  We intend to open source it within the next two weeks.

-Marcello Herreshoff

**Comment by ibab** [comment]
Thanks for posting this, @kingtaurus! This is quite useful for me.
If I'm not mistaken, your code constructs a tensorflow expression on a per-tree basis.
I think it would be more efficient to use control flow mechanisms to collect the tensors at each node and send them through a single previously defined network in batches.
But it seems that implementing it this way is quite tricky.

**Comment by kingtaurus** [comment]
@ibab - Yup, you're correct about the per-tree basis. I believe that this gives a non-ideal starting point for how to implement recursive structure in tensorflow. Right now it's problematic because the graph state has to be cleared (since each tree gets a graph, some memory may be allocated, so this has to be dealt with) and then reloaded. I think trying to shoe-horn in the other flow-control operations might be bad idea (this really doesn't fall into any of the operations like `tf.map_fn`, `tf.scan`, or `tf.foldl` or `tf.foldr`). I guess you could try using `tf.case` and `tf.while_loop` but that looks like a pain.

There are few things that need to be implemented:
(1) recursive descent graph construction (based upon 'node' type and 'loss' type). This become even more complex if you need things like attention. Ideally I would like something like,
`tf.walk( leaf_func, node_func, output_func, tree_storage, output_style="root_only | all ")`

(2) dynamic construction of and freeing of the computation graphs (you can't maintain all the possible allocated graphs in memory, but you can use flyweights and have an allocatation pool for storage).

(3) tensorboard interface/display changes to handle this new 'dynamic' graph behavior.

**Comment by ebrevdo** [comment]
You can use a while loop with TensorArray.  What you need to do is find an indexing structure mapping a linear index to nodes of the tree and back.

**Comment by ibab** [comment]
@ebrevdo: I thought about doing that at first, for example using heap indexing (or something more clever to deal with trees that are not complete).
The problem I had is that each single example in the dataset can have a different shape.
E.g. if we want to do sentiment analysis on sentences, then the number of words can be different in each sentence and the structure of each tree depends on how the corresponding sentenced was parsed.
Or maybe I've misunderstood your idea?

I think a `TensorTrees` data structure that's similar in spirit to `TensorArray` could be a nice API.
It could hold a number of trees (different tree structure for each example), and batch together all inputs/outputs that can currently be evaluated.
(Note that not all nodes inside a tree can be evaluated at the same time, as we need don't know the inputs for the upper levels of the tree at first).
Then a while loop could be used to manage the step by step evaluation of all trees.
It could work something like this (in pseudocode):
```
trees = TensorTrees()
while not trees.is_fully_evaluated():
  inputs, outputs = trees.get_batch(max_batch_size=1000)
  # Process inputs and store result in outputs
```
This would evaluate the lower levels of the trees in the first few batches, and then move upwards in later batches.
Not sure whether it would make sense to put this into the tensorflow core, maybe [tensorflow/models](https://github.com/tensorflow/models) would be a better location.

**Comment by jstaker7** [comment]
I also followed the example in CS224D until I hit a wall when trying to do mini-batching. My graph and dataset are both so large that rebuilding the graph for each tree and running them one at a time is just not feasible. I also thought about using a while_loop, but I don't see how this can work since the incoming trees have different sizes, as @ibab mentioned.

**Comment by jstaker7** [comment]
Even if we don't support a Tree datatype right away, I'd think that a way to perform recursion in-graph would be sufficient to at least allow for a working solution. Then we could arrange the trees linearly and use gather recursively to interact with the computed tensors. It looks like theano doesn't even support recursion, so perhaps this alone is a really hard problem.

**Comment by lukasr0** [comment]
I've been working with tree-structured examples on tensorflow using Control Flow ops.  While I haven't implemented the Socher et al. 2011 paper, I use the indexing structure to represent edges between nodes.  Something like:
```
  state = ta.TensorArray( tf.float32, size=1, dynamic_size=True, infer_shape=False, clear_after_read=False)

  # A super-simple tree:
  #       1   2   3
  #        \   \ /
  #         \   -
  #          \ /
  #           -
  node_features = [1.0, 2, 3]
  state = state.unpack(node_features)
  edges = [[1, 2, 3], [0, 3, 4]]

  def body(pos, state):
    edge = tf.gather(edges, pos)
    left_idx, right_idx, out_idx = tf.unpack(edge)
    # In the real-world we'd do something more useful that just tf.sub...
    out_state = tf.sub(state.read(left_idx), state.read(right_idx))
    return pos + 1, state.write(out_idx, out_state)

  def cond(pos, _):
    return tf.less(pos, len(edges))

  _, res = tf.while_loop(cond, body, [tf.constant(0), state])
  with tf.Session():
    print(res.pack().eval())
```

In order to generalize this type of processing and make it scale there is work in progress exploring support for things like:
- efficient batching - processing more than one edge at a time;
- parsing structured examples into edge and vertex tensors (kind of like tf.parse_example where one specifies in what order to traverse the tree etc.); and, 
- higher-level ops for traversing/backpropagating structured examples.  (This might well include something in the direction of what @ibab mentioned with a TensorTree.)

**Comment by ibab** [comment]
Awesome, thanks for sharing this!
I never thought of solving it this way, where the traversal instructions are "pre-packaged" by the operator that parses the examples.

**Comment by ebrevdo** [comment]
This feels like it is more appropriate as a StackOverflow thread, not github issues.

**Comment by ibab** [comment]
Here's the StackOverflow question I created originally: http://stackoverflow.com/questions/37054188/how-can-i-implement-a-recursive-neural-network-in-tensorflow

I thought that the issue might be appropriate as a feature request.
Feel free to close this issue if you prefer.

**Comment by RobRomijnders** [comment]
@ibab did you manage to implement the recursive net this way? How did it end up?

**Comment by yaroslavvb** [comment]
There's a TensorFlow library in the works that can do data-dependent computation using a single graph -- https://openreview.net/pdf?id=ryrGawqex

**Comment by cristipp** [comment]
Are there plans to open source TensorFlow Fold in the near future?

**Comment by yaroslavvb** [comment]
The paper says it'll be made available with the final version of the ICLR submission, so I guess sometime before Feb 2nd?

**Comment by kingtaurus** [comment]
@ibab I believe that this can now be solved using just `tf.while_loop` and then making sure your data is appropriately formatted. Using the idea from @bogatyy [cs224d/repo](https://github.com/bogatyy/cs224d/tree/master/assignment3) - I  modified it so that it doesn't require storage in a `tf.TensorArray` (the `while_loop` iteration is used for storage). Below is a rough sketch of a solution.

```python
    # private functions used to construct the graph.
    def _embed_word(self, word_index):
        with tf.variable_scope("Composition", reuse=True) as scope:
            embedding = tf.get_variable("embedding")
        return tf.expand_dims(tf.gather(embedding, word_index), 0)

    # private functions used to construct the graph.
    def _combine_children(self, tensor_concat, left_idx, right_idx):
        left_tensor = tf.expand_dims(tf.gather(tensor_concat, left_idx), 0)
        right_tensor = tf.expand_dims(tf.gather(tensor_concat, right_idx), 0)
        with tf.variable_scope('Composition', reuse=True):
            W1 = tf.get_variable('W1')
            b1 = tf.get_variable('b1')
        return tf.nn.relu(tf.matmul(tf.concat(1, [left_tensor, right_tensor]), W1) + b1)

    def _loop_over_tree(self, i, tensor_list):
        is_leaf = tf.gather(self.is_a_leaf, i)
        word_idx    = tf.gather(self.word_index, i)
        left_child  = tf.gather(self.left_child, i)
        right_child = tf.gather(self.right_child, i)
        node_tensor = tf.cond(is_leaf, lambda : self._embed_word(word_idx),
                                       lambda : self._combine_children(tensor_list, left_child, right_child))
        tensor_list = tf.concat(0, [tensor_list, node_tensor])
        i = tf.add(i,1)
        return i, tensor_list

    def construct_tensor_array(self):
        loop_condition = lambda i, tensor_array: \
                         tf.less(i, tf.squeeze(tf.shape(self.is_a_leaf)))
        
        left_most_element = self._embed_word(tf.gather(self.word_index, 0))
        #index should start @ 1
        i1 = tf.constant(1, dtype=tf.int32)

        while_loop_op = tf.while_loop(cond=loop_condition,
                                       body=self._loop_over_tree,
                                       loop_vars=[i1, left_most_element],
                                       shape_invariants=[i1.get_shape(), tf.TensorShape([None,50])])
        
        return while_loop_op[1]
```

where,
```python
        self.is_a_leaf   = tf.placeholder(tf.bool, [None], name="is_a_leaf")
        self.left_child  = tf.placeholder(tf.int32, [None], name="lchild")
        self.right_child = tf.placeholder(tf.int32, [None], name="rchild")
        self.word_index  = tf.placeholder(tf.int32, [None], name="word_index")
        self.labelholder = tf.placeholder(tf.int32, [None], name="labels_holder")
```

and the feed dictionary is built as follows:
```python
    def build_feed_dict(self, in_node):
        nodes_list = []
        tr.leftTraverse(in_node, lambda node, args: args.append(node), nodes_list)
        node_to_index = OrderedDict()
        for idx, i in enumerate(nodes_list):
            node_to_index[i] = idx

        feed_dict = {
          self.is_a_leaf   : [ n.isLeaf for n in nodes_list ],
          self.left_child  : [ node_to_index[n.left] if not n.isLeaf else -1 for n in nodes_list ],
          self.right_child : [ node_to_index[n.right] if not n.isLeaf else -1 for n in nodes_list ],
          self.word_index  : [ self.vocab.encode(n.word) if n.word else -1 for n in nodes_list ],
          self.labelholder : [ n.label for n in nodes_list ]
        }
        return feed_dict
```

**Comment by moshelooks** [comment]
FYI we just released [TensorFlow Fold](github.com/tensorflow/fold/). [This notebook](https://github.com/tensorflow/fold/blob/master/tensorflow_fold/g3doc/sentiment.ipynb) shows how to implement a recursive neural network.

**Comment by RobRomijnders** [comment]
Looks good. So this model assumes we know the tree a priori, doesn't it? 

So the input is (tree,words) output is (sentiment).

**Comment by RobRomijnders** [comment]
Looks good. So this implementation assumes that the tree is known a-priori, doesn't it?
So it maps (embeddings, tree) to (sentiment).

Would it also be able to map (embeddings) to (tree,sentiment)?

**Comment by moshelooks** [comment]
Great question. You're right, this example does assume that the tree structure is part of the input. The particular model in the Socher paper that generates trees can be implemented with Fold, because the shape of the computation graph is a function of the input data (I'm assuming here that the terminals and adjacency matrix are provided as inputs to the model).

However Fold in its current form will not help with arbitrary "computation-dependent-computation" e.g. a complete generative model for trees.

**Comment by RobRomijnders** [comment]
@moshelooks strong answer. thanks for the info

**Comment by girving** [comment]
I'm closing to this bug as fixed for now given Fold.  Further requests for fold features should be filed at https://github.com/tensorflow/fold.
